keps/sig-storage/20200709-pod-pvc-qos-limits.md

```diff
@@ -0,0 +1,255 @@
+* This is a k8s way, not a docker way. From cluster aspect, application owner set the pv for pods and kubelet could detect which is mount point for the pod, but which device at runtime is mounted to the pod is not visable for application owner .
+
+This is the first/alpha MVP. If this is a suggested way to implement QoS of pvc, we can add iops/bps to container limits like cpu\memory.
+
+![image](https://user-images.githubusercontent.com/2010320/88526364-72201680-d02e-11ea-8dde-891c4f96ffaf.png)
+The difference is in step 5 that kubelet will read pod annotation and add iops limits by cgroup to pod
+
+### Beta implementation of this proposal:
+
+Add the limitation directly to volumes properties like below:
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example
+  labels:
+    dce.daocloud.io/component: example
+spec:
+  containers:
+  - image: daocloud.io/daocloud/dao-2048:latest
+    name: local-2048-1
+    volumeMounts:
+    - mountPath: /data
+      name: volume
+    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
+      name: default-token-cv7mt
+      readOnly: true
+  serviceAccount: default
+  serviceAccountName: default
+  volumes:
+  - name: volume
+    persistentVolumeClaim:
+      claimName: local-storage-pvc-w-1
+    iops:
+      read: 2000
+      write: 1000
+    bps:
+      read: 1000000
+      write: 1000000
+  - name: default-token-cv7mt
+    secret:
+      defaultMode: 420
+      secretName: default-token-cv7mt
+```
+
+This would be optional properties for volume in Pod Spec. This is the beta design. 
+
+
+### API Change
+None
+
+### Implementation
+
+
+#### Implementation detail
+
+The in Kubelet creates pods. 
+This will between the volume is mounted to the host and pod is started.
+If CSI and kubelet can check which device does the pvc use, the iops limit can apply dynamicly according to PVC name.
+
+
+##### Code Design
+
+
+### User Stories
+
+1. A device is a shared volume in kubernetes clusters. We want to the write iops limitation of pod in cluster 1 is 1000 and the limitation for pod in cluster 2 to be 2000. 
+
+2. There are 100 pods that are running on the same host, among which there is a pod that uses disk heavily. Other node may be effected and we want to se an write/read iops limitation for the container on the container rootfs.
+
+
+
+other stories can be found in 
+#70364
+#27000
+#70573
+#70364
+#70980
+#92068
+
+
+### Implementation Details/Notes/Constraints
+
+
+### Risks and Mitigations
+I think iops limiting would be a great idea, but I'm not sure if the current cgroups implementation will effectively implement it. For example, with non-direct device access, writes are buffered in the kernel and something like 80% of them will not be accounted to a cgroup (instead they're all aggregated together). I did a little bit of experimentation here: https://gitlab.com/mattcary/blkio_cgroups/-/blob/master/data/blkio_cgroup.md (sorry that the writeup is not very polished).
```

Comment:
That's actually the problem with limiting IOPS: with the current Linux kernel it's not possible to limit it reliably and the reason we don't have it yet in Kubernetes.

Reply:
As docker supports it, it seems to be a general requirements for storage, especially block volumes. As I mentioned, the disk device that docker is running on, is so important that we need to protect it from being overused  by some pods.

```
--device-read-bps value       Limit read rate (bytes per second) from a device (default [])
      --device-read-iops value      Limit read rate (IO per second) from a device (default [])
      --device-write-bps value      Limit write rate (bytes per second) to a device (default [])
      --device-write-iops value     Limit write rate (IO per second) to a device (default [])

```

Reply:
if docker-shim is deprecated, I would look into other cri like containerd. Whether this is supported?

Reply:
It looks like there is some level of cgroups v2 support in containerd? https://github.com/containerd/containerd/issues/3726

containerd is supported well, it's available on GKE and is becoming the default.

https://cloud.google.com/kubernetes-engine/docs/concepts/using-containerd